renku/ui/service/gateways/repository_cache.py

```diff
@@ -0,0 +1,233 @@
+# Copyright Swiss Data Science Center (SDSC). A partnership between
+# École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Repository cache interface."""
+
+import os
+import shutil
+import uuid
+from datetime import datetime
+from typing import Optional
+from urllib.parse import urlparse
+
+import portalocker
+from marshmallow import ValidationError
+
+from renku.command.clone import project_clone_command
+from renku.core import errors
+from renku.core.util.contexts import renku_project_context
+from renku.core.util.os import normalize_to_ascii
+from renku.domain_model.git import GitURL
+from renku.infrastructure.repository import Repository
+from renku.ui.service.cache import ServiceCache
+from renku.ui.service.cache.models.project import Project
+from renku.ui.service.cache.models.user import User
+from renku.ui.service.config import PROJECT_CLONE_DEPTH_DEFAULT
+from renku.ui.service.errors import IntermittentCacheError, IntermittentLockError
+from renku.ui.service.interfaces.repository_cache import IRepositoryCache
+from renku.ui.service.logger import service_log
+
+
+class LocalRepositoryCache(IRepositoryCache):
+    """Cache for project repos stored on local disk."""
+
+    def get(
+        self, cache: ServiceCache, git_url: str, branch: Optional[str], user: User, shallow: bool = True
+    ) -> Project:
+        """Get a project from cache (clone if necessary)."""
+        if git_url is None:
+            raise ValidationError("Invalid `git_url`, URL is empty", "git_url")
+
+        try:
+            project = Project.get(
+                (Project.user_id == user.user_id) & (Project.git_url == git_url) & (Project.branch == branch)
+            )
+        except ValueError:
+            # project not found in DB
+            return self._clone_project(cache, git_url, branch, user, shallow)
+
+        if not project.abs_path.exists():
+            # cache folder doesn't exist anymore
+            project.delete()
+            return self._clone_project(cache, git_url, branch, user, shallow)
+
+        if not shallow and project.clone_depth is not None and project.clone_depth <= 1:
```

Comment:
I am assuming that clone_depth set to `0` means to actually not have a shallow clone? Maybe we should clarify that in the docs here or for the `Project`.